File: homework0/tom_nguyen/q2_FirstOccurence.cpp
#include <iostream>
#include <string>
#include <unordered_set>
using namespace std;

string firstOccurrence(string str) {
    unordered_set<char> seen;
    string result = "";
    for (char c : str) {
        if (!seen.count(c)) {
            result += c;
            seen.insert(c);
        }
    }
    return result;
}

int main() {
    cout << firstOccurrence("abracadabra") << endl;
    cout << firstOccurrence("Uber Career Prep") << endl;
    cout << firstOccurrence("zzyzx") << endl;
    return 0;
}

// Time spent: 9 minutes

File: homework0/michael_rao/q2_FirstOccurence.py
# time: O(n), space: O(n)

def first_occurence(str):
    res = ""
    seen = set()

    for c in str:
        if c not in seen:
            res += c
            seen.add(c)
    
    return res

print("Input String: abracadabra")
print("Output:", first_occurence("abracadabra"))

print("Input String: Uber Career Prep")
print("Output:", first_occurence("Uber Career Prep"))

print("Input String: aabcbc")
print("Output:", first_occurence("aabcbc"))

# took 5 minutes